### PR TITLE
feat: add zset

### PR DESCRIPTION
**Pull request by SilverRainZ**
# Intro

Package zset provides a concurrent-safety sorted set, can be used as a local replacement of Redis' zset (https://redis.com/ebook/part-2-core-concepts/chapter-3-commands-in-redis/3-5-sorted-sets/).

The main different to other sets is, every value of set is associated with a score, that is used in order to take the sorted set ordered, from the smallest to the greatest score.

The sorted set has `O(log(N))` time complexity when doing Add(ZADD) and Remove(ZREM) operations and `O(1)` time complexity when doing Contains operations.

Package zset is a go implmentation of https://github.com/redis/redis/blob/unstable/src/t_zset.c

# Comparison

| Redis command         | Go function         |
|-----------------------|---------------------|
| ZADD                  | Add                 |
| ZINCRBY               | IncrBy              |
| ZREM                  | Remove              |
| ZREMRANGEBYSCORE      | RemoveRangeByScore  |
| ZREMRANGEBYRANK       | RemoveRangeByRank   |
| ZUNION                | Union               |
| ZINTER                | Inter               |
| ZINTERCARD            | *TODO*              |
| ZDIFF                 | *TODO*              |
| ZRANGE                | Range               |
| ZRANGEBYSCORE         | IncrBy              |
| ZREVRANGEBYSCORE      | RevRangeByScore     |
| ZCOUNT                | Count               |
| ZREVRANGE             | RevRange            |
| ZCARD                 | Len                 |
| ZSCORE                | Score               |
| ZRANK                 | Rank                |
| ZREVRANK              | RevRank             |
| ZPOPMIN               | *TODO*              |
| ZPOPMAX               | *TODO*              |
| ZRANDMEMBER           | *TODO*              |

List of redis commands are generated from the following command:

```bash
cat redis/src/server.c | grep -o '"z.*",z.*Command' | grep -o '".*"' | cut -d '"' -f2
```

You may find that not all redis commands have corresponding go implementations,
the reason is as follows:

### Unsupported Commands

Redis' zset can operates elements in lexicographic order, which is not commonly
used function, so zset does not support commands like ZREMRANGEBYLEX, ZLEXCOUNT
and so on.

| Redis command         |
|-----------------------|
| ZREMRANGEBYLEX        |
| ZRANGEBYLEX           |
| ZREVRANGEBYLEX        |
| ZLEXCOUNT             |

In redis, user accesses zset via a string key. We do not need such string key
because we have variable. so the following commands are not implemented:

| Redis command         |
|-----------------------|
| ZUNIONSTORE           |
| ZINTERSTORE           |
| ZDIFFSTORE            |
| ZRANGESTORE           |
| ZMSCORE               |
| ZSCAN                 |

# Tests

## Unittests

coverage: 94.0% of statements

## Benchmark

```
go test -v -cpuprofile cpu.pprof -cpu=1 -run=NOTEST -bench=. -benchmem -benchtime=100000x -count=10 -timeout=60m > y.txt && benchstat y.txt
name                                  time/op
Contains100Hits/sortedset             39.5ns ± 1%
Contains50Hits/sortedset              43.4ns ± 4%
ContainsNoHits/sortedset              21.9ns ± 5%
Add/sortedset                         53.6ns ± 1%
1Add99Contains/sortedset              38.1ns ±12%
10Add90Contains/sortedset             47.6ns ± 5%
50Add50Contains/sortedset             56.9ns ± 3%
1Add3Incr6Remove90Contains/sortedset  49.3ns ±11%

name                                  alloc/op
Contains100Hits/sortedset              0.00B
Contains50Hits/sortedset               0.00B
ContainsNoHits/sortedset               0.00B
Add/sortedset                          2.00B ± 0%
1Add99Contains/sortedset               0.00B
10Add90Contains/sortedset              0.00B
50Add50Contains/sortedset              1.00B ± 0%
1Add3Incr6Remove90Contains/sortedset  0.50B ±100%

name                                  allocs/op
Contains100Hits/sortedset               0.00
Contains50Hits/sortedset                0.00
ContainsNoHits/sortedset                0.00
Add/sortedset                           0.00
1Add99Contains/sortedset                0.00
10Add90Contains/sortedset               0.00
50Add50Contains/sortedset               0.00
1Add3Incr6Remove90Contains/sortedset    0.00
```

# TODOs

- [x] implementation
- [x] godoc
- [x] unittest
- [x] benchmark
- [x] reamde

### CODE REVIEW

File: collection/zset/zset.go [hunk 1]
```diff
@@ -0,0 +1,879 @@
...
+// Elements with the same score are ordered lexicographically.
+//
+// This function won't panic even when the given rank out of range.
+//
+// Range is the replacement of ZRANGE command of redis.
+func (z *Float64Set) Range(start, stop int) []Float64Node {
+	z.mu.RLock()
+	defer z.mu.RUnlock()
+
+	// Convert negative rank
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:271]: One-line comment better be a full sentence.

File: collection/zset/zset.go [hunk 2]
```diff
@@ -0,0 +1,879 @@
...
+
+func (z *Float64Set) RemoveRangeByScoreWithOpt(min, max float64, opt Float64RangeOpt) []Float64Node {
+	z.mu.RLock()
+	defer z.mu.RUnlock()
+
+	return z.list.DeleteRangeByScore(min, max, opt, z.dict)
+}
+
+//
+// skip list implementation
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:416]: I am not sure maybe we could move the skiplist implementation into another file? This file is too big, more than 800 lines.

File: collection/zset/zset.go [hunk 3]
```diff
@@ -0,0 +1,879 @@
...
+//
+
+const (
+	maxLevel = 32   // Same to ZSKIPLIST_MAXLEVEL
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:420]: Comment style: better be `same to ZSKIPLIST_MAXLEVEL `

File: collection/zset/zset.go [hunk 4]
```diff
@@ -0,0 +1,879 @@
...
+
+	return removed
+}
+
+// Delete all the elements with rank between start and end from the skiplist.
+// Start and end are inclusive.
+//
+// NOTE: start and end need to be 1-based
+func (l *float64List) DeleteRangeByRank(start, end int, dict map[string]float64) []Float64Node {
+	var update [maxLevel]*float64ListNode
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:770]: Better be a declaration group

**Review by zhangyunhao116 — COMMENTED** [review]
Maybe we could support `NewInt` in this pull request? (In next PR is also ok)

File: collection/zset/zset.go [hunk 5]
```diff
@@ -0,0 +1,879 @@
...
+import (
+	"math"
+	"sync"
+	"unsafe"
+
+	"github.com/bytedance/gopkg/lang/fastrand"
+)
+
+// Float64RangeOpt describes the whether the min/max is exclusive in score range.
+type Float64RangeOpt struct {
```
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/zset.go:34]: We may generate code for other score types, how can we prevent the conflicting of `RangeOpt` ?

File: collection/zset/zset.go (hunk 1 above)
Comment (SilverRainZ) [reply collection/zset/zset.go]: Done.

**Comment by SilverRainZ** [comment]
> Maybe we could support `NewInt` in this pull request? (In next PR is also ok)

I prefer in next PR, just a code generateion stuff, we focus on the core implementaion in this PR.

File: collection/zset/zset.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhangyunhao116) [reply collection/zset/zset.go:34]: Probably we could move `RangeOpt` into another file, which makes it available for all types.

File: collection/zset/zset.go (hunk 2 above)
Comment (zhangyunhao116) [reply collection/zset/zset.go]: Yes, maybe move all reusable parts into other files is a better idea.

File: collection/zset/zset.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhangyunhao116) [reply collection/zset/zset.go:34]: 因为 `RangeOpt` 其实和类型无关吧，里面就是两个 bool 值，代表区间的开关。无论是 IntRangeOpt 还是 FloatRangeOpt 都是两个 bool 值，所以复用一个 struct 就好啦

File: collection/zset/zset.go [hunk 6]
```diff
@@ -0,0 +1,885 @@
...
+		update[i] = x
+	}
+
+	// We assume the element is not already inside, since we allow duplicated
+	// scores, reinserting the same element should never happen since the
+	// caller of Add() should test in the hash table if the element is
+	// already inside or not.
+	level := l.randomLevel()
+	if level > l.highestLevel {
+		// create higher levels
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:543]:
```suggestion
		// Create higher levels.
```

File: collection/zset/zset.go [hunk 7]
```diff
@@ -0,0 +1,885 @@
...
+	x = newFloat64ListNode(score, value, level)
+	for i := 0; i < level; i++ {
+		// update --> x --> update.next
+		x.storeNext(i, update[i].loadNext(i))
+		update[i].storeNext(i, x)
+		// update[i].span is splitted to: new update[i].span and x.span
+		x.storeSpan(i, update[i].loadSpan(i)-(rank[0]-rank[i]))
+		update[i].storeSpan(i, (rank[0]-rank[i])+1)
+	}
+	// Increment span for untouched levels
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:560]:
```suggestion
	// Increment span for untouched levels.
```

File: collection/zset/zset.go [hunk 8]
```diff
@@ -0,0 +1,885 @@
...
+	for i := level; i < l.highestLevel; i++ {
+		update[i].storeSpan(i, update[i].loadSpan(i)+1)
+	}
+
+	// update back pointer
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:565]:
```suggestion
	// Update back pointer.
```

File: collection/zset/zset.go [hunk 9]
```diff
@@ -0,0 +1,885 @@
...
+		next.prev = x
+	} else {
+		l.tail = x
+	}
+	l.length++
+
+	return x
+}
+
+// randomLevel returns a level between (1, maxLevel] for insertion.
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:580]:
```suggestion
// randomLevel returns a level between [1, maxLevel] for insertion.
```

File: collection/zset/zset.go [hunk 10]
```diff
@@ -0,0 +1,885 @@
...
+// update position matrix.
+func (l *float64List) deleteNode(x *float64ListNode, update *[maxLevel]*float64ListNode) {
+	for i := 0; i < l.highestLevel; i++ {
+		if update[i].loadNext(i) == x {
+			// Remove x, updaet[i].span = updaet[i].span + x.span - 1 (x removed)
+			next, span := x.loadNextAndSpan(i)
+			span += update[i].loadSpan(i) - 1
+			update[i].storeNextAndSpan(i, next, span)
+		} else {
+			// x does not appear on this level, just update span
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:627]:
```suggestion
			// x does not appear on this level, just update span.
```

File: collection/zset/zset.go [hunk 11]
```diff
@@ -0,0 +1,885 @@
...
+			update[i].storeSpan(i, update[i].loadSpan(i)-1)
+		}
+	}
+	if next := x.loadNext(0); next != nil { // not tail of skiplist
+		next.prev = x.prev
+	} else {
+		l.tail = x.prev
+	}
+	for l.highestLevel > 1 && l.header.loadNext(l.highestLevel-1) != nil {
+		// Clear the pointer and span for safety
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:637]:
```suggestion
		// Clear the pointer and span for safety.
```

File: collection/zset/zset.go [hunk 12]
```diff
@@ -0,0 +1,885 @@
...
+			next = x.loadNext(i)
+		}
+		update[i] = x
+	}
+	x = x.loadNext(0)
+	if x != nil && x.equal(score, value) {
+		l.deleteNode(x, &update)
+		return x
+	}
+	// not found
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:663]: Looks like redundant comment.
```suggestion
```

File: collection/zset/zset.go [hunk 13]
```diff
@@ -0,0 +1,885 @@
...
+	for i := l.highestLevel - 1; i >= 0; i-- {
+		next := x.loadNext(i)
+		for next != nil && !greaterThanMin(next.score, min, opt.ExcludeMin) {
+			x = next
+			next = x.loadNext(i)
+		}
+		update[i] = x
+	}
+
+	// Current node is the last with score not greater than min. */
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:751]:
```suggestion
	// Current node is the last with score not greater than min.
```

File: collection/zset/zset.go [hunk 14]
```diff
@@ -0,0 +1,885 @@
...
+	x = x.loadNext(0)
+
+	// Delete nodes in range
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:754]:
```suggestion
	// Delete nodes in range.
```

File: collection/zset/zset.go [hunk 15]
```diff
@@ -0,0 +1,885 @@
...
+			traversed += span
+			x = next
+			next, span = x.loadNextAndSpan(i)
+		}
+		update[i] = x
+	}
+
+	traversed++
+	x = x.loadNext(0)
+	// Delete nodes in range
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:793]:
```suggestion
	// Delete nodes in range.
```

File: collection/zset/zset.go [hunk 16]
```diff
@@ -0,0 +1,885 @@
...
+	x := l.header
+	for i := l.highestLevel - 1; i >= 0; i-- {
+		next := x.loadNext(i)
+		for next != nil && !greaterThanMin(next.score, min, opt.ExcludeMin) {
+			x = next
+			next = x.loadNext(i)
+		}
+	}
+
+	// The next node MUST not be NULL (excluded by IsInRange)
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:842]:
```suggestion
	// The next node MUST not be NULL (excluded by IsInRange).
```

File: collection/zset/zset.go [hunk 17]
```diff
@@ -0,0 +1,885 @@
...
+	x := l.header
+	for i := l.highestLevel - 1; i >= 0; i-- {
+		next := x.loadNext(i)
+		for next != nil && lessThanMax(next.score, max, opt.ExcludeMax) {
+			x = next
+			next = x.loadNext(i)
+		}
+	}
+
+	// The node x must not be NULL (excluded by IsInRange)
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:865]:
```suggestion
	// The node x must not be NULL (excluded by IsInRange).
```

File: collection/zset/zset.go [hunk 18]
```diff
@@ -0,0 +1,885 @@
...
+	if !greaterThanMin(x.score, min, opt.ExcludeMin) {
+		return nil
+	}
+	return x
+}
+
+// IsInRange returns whether there is a port of sorted set in given range.
+func (l *float64List) IsInRange(min, max float64, opt Float64RangeOpt) bool {
+	// Test empty range
```
Comment (zhangyunhao116) [inline collection/zset/zset.go:874]:
```suggestion
	// Test empty range.
```

**Review by zhangyunhao116 — COMMENTED** [review]
一些风格问题，统一一下就好啦，已经加了 suggestion

File: collection/zset/zset.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/zset.go:34]: 啊对的对的……糊涂了。

File: collection/zset/readme.md [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+|-----------------------|---------------------|
+| ZADD                  | Add                 |
+| ZINCRBY               | IncrBy              |
+| ZREM                  | Remove              |
+| ZREMRANGEBYSCORE      | RemoveRangeByScore  |
+| ZREMRANGEBYRANK       | RemoveRangeByRank   |
+| ZREMRANGEBYLEX        | NOT SUPPORTED       |
+| ZUNIONSTORE           | UnionStore          |
+| ZINTERSTORE           | InterStore          |
+| ZDIFFSTORE            |                     |
```
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/readme.md:31]: 在内部的需求文档里没覆盖到，因此这次先不实现，以后可能会实现。

File: collection/zset/skiplist.go [hunk 20]
```diff
@@ -0,0 +1,492 @@
...
+	"github.com/bytedance/gopkg/lang/fastrand"
+)
+
+//
+// Skip list implementation.
+//
+
+const (
+	maxLevel = 32   // same to ZSKIPLIST_MAXLEVEL
+	p        = 0.25 // same to ZSKIPLIST_P
```
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/skiplist.go:29]: 这里是保持和 redis 的常量名一致，也许后面加点注释会好一点？

File: collection/zset/oparry.go [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+// limitations under the License.
+
+package zset
+
+import (
+	"unsafe"
+)
+
+const (
+	op1 = 4
```
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/oparry.go:22]: 这里是和 skipset/skipmap 里的 oparray 保持一致，它们都是同一份 codebase 但没法共享。

File: collection/zset/skiplist.go [hunk 22]
```diff
@@ -0,0 +1,492 @@
...
+	oparr optionalArray
+}
+
+func newFloat64ListNode(score float64, value string, level int) *float64ListNode {
+	node := &float64ListNode{
+		score: score,
+		value: value,
+		level: level,
+	}
+	if level > op1 {
```
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/skiplist.go:47]: 是的，之前想这么写，但处于保持一致的原因没有这么做…

File: collection/zset/zset.go [hunk 23]
```diff
@@ -0,0 +1,403 @@
...
+}
+
+// UnionStoreFloat64 returns the union of given sorted sets, the resulting score of
+// a value is the sum of its scores in the sorted sets where it exists.
+//
+// UnionStoreFloat64 is the replacement of UNIONSTORE command of redis.
+func UnionStoreFloat64(zs ...*Float64Set) *Float64Set {
+	dest := NewFloat64()
+	for _, z := range zs {
+		for _, n := range z.Range(0, z.Len()-1) {
```
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/zset.go:58]: 有道理！确实存在问题。我用 `z.Range(0, -1)` 可以在同个锁内获取 len，我改一下。

File: collection/zset/zset.go [hunk 24]
```diff
@@ -0,0 +1,403 @@
...
+			}
+		}
+		if ok {
+			dest.Add(n.Score, n.Value)
+		}
+	}
+	return dest
+}
+
+// Len returns the length of Float64Set.
```
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/zset.go:88]: 这里我觉得还是规范重要些，`Len()` 在 go 里面常用得多，map 也用了 len。

File: collection/zset/readme.md [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+zset provides a concurrent-safety sorted set, can be used as a local replacement of [Redis' zset](https://redis.com/ebook/part-2-core-concepts/chapter-3-commands-in-redis/3-5-sorted-sets/).
+
+The main different to other sets is, every value of set is associated with a score, that is used in order to take the sorted set ordered, from the smallest to the greatest score.
+
+The sorted set has `O(log(N))` time complexity when doing Add(ZADD) and Remove(ZREM) operations and `O(1)` time complexity when doing Contains operations.
+
+## Features
+
+- Concurrent safe API
+- Value is sorted with score
```
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/readme.md:16]: 我理解这就是元素按分数排序就是对 zset 的主要需求了？

实现上应该说用 skiplist blahblah……

File: collection/zset/skiplist.go [hunk 26]
```diff
@@ -0,0 +1,492 @@
...
+
+import (
+	"math"
+	"unsafe"
+
+	"github.com/bytedance/gopkg/lang/fastrand"
+)
+
+//
+// Skip list implementation.
```
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/skiplist.go:24]: 这一点问过 @zhangyunhao116 了，结论是不能：skipset（以及标准的 skipllist）的底层数据结构是单链表的，而 zset 中的 skiplist 为了实现逆序遍历，底层是个双向链表。

File: collection/zset/readme.md [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+You may find that not all redis commands have corresponding go implementations,
+the reason is as follows:
+
+### Not going to support LEX commands
+
+Redis' zset can operates elements in lexicographic order, which is not commonly
+used function, so zset does not support commands like ZREMRANGEBYLEX, ZLEXCOUNT
+and so on.
+
+### Not implemented yet
```
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/readme.md]: removed.

File: collection/zset/readme.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (luchsh) [reply collection/zset/readme.md:31]: `<NOT SUPPORTED>, <TODO>`

[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/readme.md:31]: Done, 用斜体区分了，`<>` 在 markdown 里不友好。

File: collection/zset/skiplist.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/skiplist.go:29]: 改成 `probability` 了。

File: collection/zset/skiplist.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/skiplist.go:47]: done。

File: collection/zset/zset.go [hunk 28]
```diff
@@ -0,0 +1,403 @@
...
+func InterStoreFloat64(zs ...*Float64Set) *Float64Set {
+	dest := NewFloat64()
+	if len(zs) == 0 {
+		return dest
+	}
+	for _, n := range zs[0].Range(0, -1) {
+		ok := true
+		for _, z := range zs[1:] {
+			if !z.Contains(n.Value) {
+				ok = false
```
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/zset.go:78]: fixed

File: collection/zset/readme.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/readme.md:16]: 只改了语法……

File: collection/zset/readme.md [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+| Redis command         | Go function         |
+|-----------------------|---------------------|
+| ZADD                  | Add                 |
+| ZINCRBY               | IncrBy              |
+| ZREM                  | Remove              |
+| ZREMRANGEBYSCORE      | RemoveRangeByScore  |
+| ZREMRANGEBYRANK       | RemoveRangeByRank   |
+| ZREMRANGEBYLEX        | *NOT SUPPORTED*     |
+| ZUNIONSTORE           | *NOT SUITABLE*      |
+| ZINTERSTORE           | *NOT SUITABLE*      |
```
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/readme.md:31]: 重新整理了格式，把不打算支持的命令分开放了。

**Review by luchsh — DISMISSED** [review]

File: collection/zset/readme.md [hunk 30]
```diff
@@ -0,0 +1,112 @@
...
+
+[![Go Reference](https://pkg.go.dev/badge/github.com/bytedance/gopkg/collection/zset.svg)](https://pkg.go.dev/github.com/bytedance/gopkg/collection/zset)
+
+## Introduction
+
+zset provides a concurrent-safety sorted set, can be used as a local replacement of [Redis' zset](https://redis.com/ebook/part-2-core-concepts/chapter-3-commands-in-redis/3-5-sorted-sets/).
+
+The main different to other sets is, every value of set is associated with a score, that is used in order to take the sorted set ordered, from the smallest to the greatest score.
+
+The sorted set has `O(log(N))` time complexity when doing Add(ZADD) and Remove(ZREM) operations and `O(1)` time complexity when doing Contains operations.
```
Comment (zhangyunhao116) [inline collection/zset/readme.md:11]: Maybe `The sorted set`  -> `The zset` ? It took me a moment to think about what is going on here haha.

File: collection/zset/readme.md [hunk 31]
```diff
@@ -0,0 +1,112 @@
+# zset
+
+[![Go Reference](https://pkg.go.dev/badge/github.com/bytedance/gopkg/collection/zset.svg)](https://pkg.go.dev/github.com/bytedance/gopkg/collection/zset)
+
+## Introduction
+
+zset provides a concurrent-safety sorted set, can be used as a local replacement of [Redis' zset](https://redis.com/ebook/part-2-core-concepts/chapter-3-commands-in-redis/3-5-sorted-sets/).
+
+The main different to other sets is, every value of set is associated with a score, that is used in order to take the sorted set ordered, from the smallest to the greatest score.
```
[the remark this answers is not in the document]
Comment (SilverRainZ) [reply collection/zset/readme.md]: Corrected :D

File: collection/zset/readme.md (hunk 30 above)
Comment (SilverRainZ) [reply collection/zset/readme.md]: Yes, "zset" is bettter, corrected :D

**Review by zhangyunhao116 — DISMISSED** [review]

**Review by luchsh — APPROVED** [review]

**Review by luchsh — APPROVED** [review]

**Review by zhangyunhao116 — APPROVED** [review]

**Pull request merged**